### PR TITLE
refactor(resize): update for v3

### PR DESCRIPTION
**Pull request by sh7dm**
## Description
Refactor resize directive for v3.

## Motivation and Context
V3 update

## How Has This Been Tested?
`jest`

## Markup
<details>

```vue
<script>
import { Resize } from '../src/directives/resize'
import { defineComponent, withDirectives, h } from 'vue'

export default defineComponent(() => () => withDirectives(h('div', { class: 'test' }), [ [ Resize, () => console.log('resize'), "", { quiet: true } ] ]))
</script>

<style>
  .test {
    width: 100px;
    height: 100px;
    background-color: red;
  }
</style>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/directives/resize/index.ts [hunk 1]
```diff
@@ -1,11 +1,11 @@
-import { VNodeDirective } from 'vue/types/vnode'
+import { DirectiveBinding, ObjectDirective } from 'vue'
 
-interface ResizeVNodeDirective extends VNodeDirective {
-  value?: () => void
+interface ResizeVNodeDirective extends DirectiveBinding {
```
Comment (sh7dm) [inline packages/vuetify/src/directives/resize/index.ts]: renaming this:
```suggestion
interface ResizeDirectiveBinding extends DirectiveBinding {
```

File: packages/vuetify/src/directives/resize/index.ts [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
+  value: () => void | undefined
   options?: boolean | AddEventListenerOptions
 }
 
-function inserted (el: HTMLElement, binding: ResizeVNodeDirective) {
+function mounted (el: HTMLElement, binding: ResizeVNodeDirective) {
```
Comment (sh7dm) [inline packages/vuetify/src/directives/resize/index.ts]: renaming this:
```suggestion
function mounted (el: HTMLElement, binding: ResizeDirectiveBinding) {
```

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 3]
```diff
@@ -1,30 +1,33 @@
 // Directives
 import Resize from '../'
+import { h } from 'vue';
```
Comment (sh7dm) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion

// Libraries
import { h } from 'vue'
```

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 4]
```diff
@@ -1,30 +1,35 @@
 // Directives
 import Resize from '../'
 
+// Libraries
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]: move before directive import

File: packages/vuetify/src/directives/resize/index.ts [hunk 5]
```diff
@@ -1,11 +1,11 @@
-import { VNodeDirective } from 'vue/types/vnode'
+import { DirectiveBinding, ObjectDirective } from 'vue'
 
-interface ResizeVNodeDirective extends VNodeDirective {
-  value?: () => void
+interface ResizeDirectiveBinding extends DirectiveBinding {
+  value: () => void | undefined
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/directives/resize/index.ts]: why is `undefined` necessary?

Comment (sh7dm) [reply packages/vuetify/src/directives/resize/index.ts]: value can't have `?` in Vue 3.

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 6]
```diff
@@ -1,3 +1,6 @@
...
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion
// Setup
```

File: packages/vuetify/src/directives/resize/index.ts (hunk 5 above)
Comment (jacekkarczmarczyk) [reply packages/vuetify/src/directives/resize/index.ts]: Ah, I thought it's fn returning void or undefined, maybe a good idea to add parens here?

Comment (sh7dm) [reply packages/vuetify/src/directives/resize/index.ts]: will check using ts playground. I'm unsure about the order here.

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/directives/resize/index.ts]: Even if it works parens could be added to avoid confusion

Comment (sh7dm) [reply packages/vuetify/src/directives/resize/index.ts]: Thanks! Parens here are must-have things!

image

```suggestion
  value: (() => void) | undefined
```

**Comment by nekosaur** [comment]
playground?

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 7]
```diff
@@ -6,25 +9,27 @@ describe('resize.ts', () => {
...
     const el = {}
 
-    Resize.inserted(el as HTMLElement, { value: callback } as any)
+    Resize.mounted(el as HTMLElement, { value: callback } as any, vnode, null)
     expect(callback).toHaveBeenCalled()
     expect(window.addEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
-    Resize.unbind(el as HTMLElement)
+    Resize.unmounted(el as HTMLElement, {} as any, vnode, vnode)
     expect(window.removeEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
   })
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]: clear mocks after test (see https://github.com/vuetifyjs/vuetify/pull/10447/files#diff-f3e43c96b7a438392c87429ac25fe5f4R51)

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 8]
```diff
@@ -6,25 +9,27 @@ describe('resize.ts', () => {
...
     const el = {}
 
-    Resize.inserted(el as HTMLElement, { value: callback, modifiers: { quiet: true } } as any)
+    Resize.mounted(el as HTMLElement, { value: callback, modifiers: { quiet: true } } as any, vnode, null)
     expect(callback).not.toHaveBeenCalled()
     expect(window.addEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
-    Resize.unbind(el as HTMLElement)
+    Resize.unmounted(el as HTMLElement, {} as any, vnode, vnode)
     expect(window.removeEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
   })
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]: clear mocks after test (see https://github.com/vuetifyjs/vuetify/pull/10447/files#diff-f3e43c96b7a438392c87429ac25fe5f4R51)

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 9]
```diff
@@ -6,25 +9,31 @@ describe('resize.ts', () => {
     const callback = jest.fn()
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion
    const callback = jest.fn()

```

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 10]
```diff
@@ -6,25 +9,31 @@ describe('resize.ts', () => {
...
     jest.spyOn(window, 'addEventListener')
     jest.spyOn(window, 'removeEventListener')
+    const vnode = h('div')
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion

    const vnode = h('div')
```

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 11]
```diff
@@ -6,25 +9,31 @@ describe('resize.ts', () => {
...
     const el = {}
 
-    Resize.inserted(el as HTMLElement, { value: callback } as any)
+    Resize.mounted(el as HTMLElement, { value: callback } as any, vnode, null)
     expect(callback).toHaveBeenCalled()
     expect(window.addEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
-    Resize.unbind(el as HTMLElement)
+    Resize.unmounted(el as HTMLElement, {} as any, vnode, vnode)
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion

    Resize.unmounted(el as HTMLElement, {} as any, vnode, vnode)
```

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 12]
```diff
@@ -6,25 +9,31 @@ describe('resize.ts', () => {
...
     expect(window.removeEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
+    ;(window.addEventListener as jest.Mock).mockClear()
+    ;(window.removeEventListener as jest.Mock).mockClear()
   })
 
   it('should not run the callback in quiet mode', () => {
     const callback = jest.fn()
     jest.spyOn(window, 'addEventListener')
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion

    jest.spyOn(window, 'addEventListener')
```

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 13]
```diff
@@ -6,25 +9,31 @@ describe('resize.ts', () => {
...
     jest.spyOn(window, 'removeEventListener')
+    const vnode = h('div')
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion

    const vnode = h('div')
```

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 14]
```diff
@@ -6,25 +9,31 @@ describe('resize.ts', () => {
...
     const el = {}
 
-    Resize.inserted(el as HTMLElement, { value: callback, modifiers: { quiet: true } } as any)
+    Resize.mounted(el as HTMLElement, { value: callback, modifiers: { quiet: true } } as any, vnode, null)
     expect(callback).not.toHaveBeenCalled()
     expect(window.addEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
-    Resize.unbind(el as HTMLElement)
+    Resize.unmounted(el as HTMLElement, {} as any, vnode, vnode)
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion

    Resize.unmounted(el as HTMLElement, {} as any, vnode, vnode)
```

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 15]
```diff
@@ -6,25 +9,31 @@ describe('resize.ts', () => {
...
     expect(window.removeEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
+    ;(window.addEventListener as jest.Mock).mockClear()
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion

    ;(window.addEventListener as jest.Mock).mockClear()
```

File: packages/vuetify/src/directives/resize/__tests__/resize.spec.ts [hunk 16]
```diff
@@ -6,25 +9,31 @@ describe('resize.ts', () => {
...
     const el = {}
 
-    Resize.inserted(el as HTMLElement, { value: callback } as any)
+    Resize.mounted(el as HTMLElement, { value: callback } as any, vnode, null)
     expect(callback).toHaveBeenCalled()
     expect(window.addEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
-    Resize.unbind(el as HTMLElement)
+    Resize.unmounted(el as HTMLElement, {} as any, vnode, vnode)
     expect(window.removeEventListener).toHaveBeenCalledWith('resize', callback, { passive: true })
+    ;(window.addEventListener as jest.Mock).mockClear()
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/__tests__/resize.spec.ts]:
```suggestion

    ;(window.addEventListener as jest.Mock).mockClear()
```

File: packages/vuetify/src/directives/resize/index.ts [hunk 17]
```diff
@@ -1,13 +1,14 @@
...
+  value: (() => void) | undefined
 }
 
-function inserted (el: HTMLElement, binding: ResizeVNodeDirective) {
+function mounted (el: HTMLElement, binding: ResizeDirectiveBinding) {
   const callback = binding.value!
-  const options = binding.options || { passive: true }
+  const options = {
+    passive: !((binding.modifiers && binding.modifiers.active) || false)
+  } as AddEventListenerOptions
```
Comment (johnleider) [inline packages/vuetify/src/directives/resize/index.ts]:
```suggestion
  } as AddEventListenerOptions

```

File: packages/vuetify/src/directives/resize/index.ts [hunk 18]
```diff
@@ -1,13 +1,15 @@
...
-  options?: boolean | AddEventListenerOptions
+interface ResizeDirectiveBinding extends DirectiveBinding {
+  value: (() => void) | undefined
 }
 
-function inserted (el: HTMLElement, binding: ResizeVNodeDirective) {
+function mounted (el: HTMLElement, binding: ResizeDirectiveBinding) {
   const callback = binding.value!
-  const options = binding.options || { passive: true }
+  const options = {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/directives/resize/index.ts]: how about (ts.next version)
```ts
const options = binding.modifiers?.active 
  ? (binding.options ?? false)
  : passiveEventOptions(binding.options ?? false)
```

Comment (sh7dm) [reply packages/vuetify/src/directives/resize/index.ts]: maybe. do we have a version that supports it?

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/directives/resize/index.ts]: Not yet, although you can skip the ts.next part for now and apply the rest of the proposal

```ts
const options = (binding.modifiers && binding.modifiers.active) 
  ? (binding.options || false)
  : passiveEventOptions(binding.options || false)
```

Comment (sh7dm) [reply packages/vuetify/src/directives/resize/index.ts]: sure! could you create a suggestion please?

Comment (KaelWD) [reply packages/vuetify/src/directives/resize/index.ts]: We aren't supporting IE so PASSIVE_SUPPORTED can be assumed to be true

**Review by nekosaur — APPROVED** [review]

**Pull request merged**